crates/ruff_server/src/session/index.rs

```diff
@@ -310,10 +309,8 @@ impl Index {
             return;
         };
 
-        // TODO: I think this does not correctly reload settings when using `extend` and the extended
-        //  setting isn't in a parent folder.
-        for (root, settings) in self.settings.range_mut(enclosing_folder.to_path_buf()..) {
-            if !root.starts_with(enclosing_folder) {
+        for (root, settings) in &mut self.settings {
+            if !enclosing_folder.starts_with(root) {
```

Comment:
There are two main changes:
1. We loop over every workspace folder. The `range_mut` is incorrect here because the "enclosing_folder" would be within the workspace.
2. The `starts_with` call is incorrect. We should be checking if the folder is part of the workspace or not and not the other way around which will always be incorrect for nested folders.

Reply:
I wonder if this is correct though: Let's say we have workspace `a` and `b` and we add a configuration to `b/foo` so that `root` = `a` and `enclosing_folder=b/foo`. We would immediately hit the `break` here. 

I think we want `self.settings.range(..enclosing_folder).rev()` and we take the paths for as long as they start with `enclosing_folder`.

Reply:
Oh, that's correct. Thanks. Let me update it.

Reply:
This is actually not correct either because the `root` folder can never start with an `enclosing_folder`. This is the main reason I switched the `starts_with` call in (2). Let me think about this a bit more considering your example as well.